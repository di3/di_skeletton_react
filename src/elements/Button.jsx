import React, { Component } from 'react';
import { C1, C2, C3, C4 } from '../style/ci/color';
import disableTextSelectionStyle from '../style/disableTextSelection';
import m from 'di_merge';

class Button extends Component {
  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
    this.handleMouseOver = this.handleMouseOver.bind(this);
    this.handleMouseLeave = this.handleMouseLeave.bind(this);
    this.state = {hover: false};
  }

  handleMouseOver() {
    this.setState({hover: true});
  }

  handleMouseLeave() {
    this.setState({hover: false});
  }

  handleClick() {
    this.props.onClick();
  }

  getContainerStyle(style = {}) {
    const { hover } = this.state;
    return m({
      fontFamily: "Monospace",
      display:"block",
      margin: "auto",
      fontSize: "180%",
      border: "4px solid " + (hover ? C1 : C2),
      backgroundColor: hover ? C3 : C4,
      color: hover ? C1 : C2,
      cursor: hover ? "pointer" : null
    }, disableTextSelectionStyle, style);
  }

  render() {
    const { children, style } = this.props;
    return (
      <div
        style={this.getContainerStyle(style)}
        onClick={this.handleClick}
        onMouseOver={this.handleMouseOver}
        onMouseLeave={this.handleMouseLeave}
      >
        {children}
      </div>
    );
  }
}

export default Button;
