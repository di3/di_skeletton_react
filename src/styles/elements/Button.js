import m from 'di_get_merge';
import { C1, C2, C3, C4 } from '../ci/color';
import disableTextSelectionStyle from '../disableTextSelection';

export const getContainerStyle = (props, state) => {
  const { hover } = state;
  const { style } = props;
  return m({
    fontFamily: "Monospace",
    display:"block",
    margin: "auto",
    fontSize: "180%",
    border: "4px solid " + (hover ? C1 : C2),
    backgroundColor: hover ? C3 : C4,
    color: hover ? C1 : C2,
    cursor: hover ? "pointer" : null
  }, disableTextSelectionStyle, style);
}
