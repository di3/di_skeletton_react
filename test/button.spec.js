import React from 'react';
import { render } from 'react-dom';
import { shallow } from 'enzyme';
import Button from 'Button';

describe('Button', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    render(<Button>reset</Button>, div);
  });
  describe('render', () => {
    it('should render the button', () => {
      const button = shallow(<Button>reset</Button>);
      expect(button.text()).toEqual('reset');
    });
    it('should handle click', () => {
      var clicked = false;
      const button = shallow(<Button onClick={() => {clicked = true}}>reset</Button>);
      button.simulate("click");
      expect(clicked).toBe(true);
    });
    it('should handle hover', () => {
      const button = shallow(<Button onClick={() => {clicked = true}}>reset</Button>);
      expect(button).toMatchSnapshot();
      button.simulate("mouseOver");
      expect(button).toMatchSnapshot();
      button.simulate("mouseleave");
      expect(button).toMatchSnapshot();
    });
  });
});
