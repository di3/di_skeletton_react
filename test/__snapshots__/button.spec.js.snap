// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button render should handle hover 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    onClick={[Function]}
  >
    reset
  </Button>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "reset",
      "onClick": [Function],
      "onMouseLeave": [Function],
      "onMouseOver": [Function],
      "style": Object {
        "KhtmlUserSelect": "none",
        "MozUserSelect": "none",
        "OUserSelect": "none",
        "WebkitTouchCallout": "none",
        "WebkitUserSelect": "none",
        "backgroundColor": "#966f8f",
        "border": "4px solid #562f4f",
        "color": "#562f4f",
        "cursor": null,
        "display": "block",
        "fontFamily": "Monospace",
        "fontSize": "180%",
        "margin": "auto",
        "msUserSelect": "none",
        "userSelect": "none",
      },
    },
    "ref": null,
    "rendered": "reset",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "reset",
        "onClick": [Function],
        "onMouseLeave": [Function],
        "onMouseOver": [Function],
        "style": Object {
          "KhtmlUserSelect": "none",
          "MozUserSelect": "none",
          "OUserSelect": "none",
          "WebkitTouchCallout": "none",
          "WebkitUserSelect": "none",
          "backgroundColor": "#966f8f",
          "border": "4px solid #562f4f",
          "color": "#562f4f",
          "cursor": null,
          "display": "block",
          "fontFamily": "Monospace",
          "fontSize": "180%",
          "margin": "auto",
          "msUserSelect": "none",
          "userSelect": "none",
        },
      },
      "ref": null,
      "rendered": "reset",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Button render should handle hover 2`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    onClick={[Function]}
  >
    reset
  </Button>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "reset",
      "onClick": [Function],
      "onMouseLeave": [Function],
      "onMouseOver": [Function],
      "style": Object {
        "KhtmlUserSelect": "none",
        "MozUserSelect": "none",
        "OUserSelect": "none",
        "WebkitTouchCallout": "none",
        "WebkitUserSelect": "none",
        "backgroundColor": "#4f6f76",
        "border": "4px solid #2f4f56",
        "color": "#2f4f56",
        "cursor": "pointer",
        "display": "block",
        "fontFamily": "Monospace",
        "fontSize": "180%",
        "margin": "auto",
        "msUserSelect": "none",
        "userSelect": "none",
      },
    },
    "ref": null,
    "rendered": "reset",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "reset",
        "onClick": [Function],
        "onMouseLeave": [Function],
        "onMouseOver": [Function],
        "style": Object {
          "KhtmlUserSelect": "none",
          "MozUserSelect": "none",
          "OUserSelect": "none",
          "WebkitTouchCallout": "none",
          "WebkitUserSelect": "none",
          "backgroundColor": "#4f6f76",
          "border": "4px solid #2f4f56",
          "color": "#2f4f56",
          "cursor": "pointer",
          "display": "block",
          "fontFamily": "Monospace",
          "fontSize": "180%",
          "margin": "auto",
          "msUserSelect": "none",
          "userSelect": "none",
        },
      },
      "ref": null,
      "rendered": "reset",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Button render should handle hover 3`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    onClick={[Function]}
  >
    reset
  </Button>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "reset",
      "onClick": [Function],
      "onMouseLeave": [Function],
      "onMouseOver": [Function],
      "style": Object {
        "KhtmlUserSelect": "none",
        "MozUserSelect": "none",
        "OUserSelect": "none",
        "WebkitTouchCallout": "none",
        "WebkitUserSelect": "none",
        "backgroundColor": "#966f8f",
        "border": "4px solid #562f4f",
        "color": "#562f4f",
        "cursor": null,
        "display": "block",
        "fontFamily": "Monospace",
        "fontSize": "180%",
        "margin": "auto",
        "msUserSelect": "none",
        "userSelect": "none",
      },
    },
    "ref": null,
    "rendered": "reset",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "reset",
        "onClick": [Function],
        "onMouseLeave": [Function],
        "onMouseOver": [Function],
        "style": Object {
          "KhtmlUserSelect": "none",
          "MozUserSelect": "none",
          "OUserSelect": "none",
          "WebkitTouchCallout": "none",
          "WebkitUserSelect": "none",
          "backgroundColor": "#966f8f",
          "border": "4px solid #562f4f",
          "color": "#562f4f",
          "cursor": null,
          "display": "block",
          "fontFamily": "Monospace",
          "fontSize": "180%",
          "margin": "auto",
          "msUserSelect": "none",
          "userSelect": "none",
        },
      },
      "ref": null,
      "rendered": "reset",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
